File: android/app/src/main/java/com/whisper2/app/services/auth/AuthService.kt
package com.whisper2.app.services.auth

import android.util.Base64
import com.google.firebase.messaging.FirebaseMessaging
import com.google.gson.Gson
import com.google.gson.JsonElement
import com.whisper2.app.core.Constants
import com.whisper2.app.core.Logger
import com.whisper2.app.crypto.CryptoService
import com.whisper2.app.data.local.prefs.SecureStorage
import com.whisper2.app.data.network.ws.ErrorPayload
import com.whisper2.app.data.network.ws.LogoutPayload
import com.whisper2.app.data.network.ws.RegisterAckPayload
import com.whisper2.app.data.network.ws.RegisterBeginPayload
import com.whisper2.app.data.network.ws.RegisterChallengePayload
import com.whisper2.app.data.network.ws.RegisterProofPayload
import com.whisper2.app.data.network.ws.SessionRefreshAckPayload
import com.whisper2.app.data.network.ws.UpdateTokensPayload
import com.whisper2.app.data.network.ws.WsClientImpl
import com.whisper2.app.data.network.ws.WsConnectionState
import com.whisper2.app.data.network.ws.WsFrame
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.tasks.await
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

sealed class AuthState {
    object Unauthenticated : AuthState()
    object Authenticating : AuthState()
    data class Authenticated(val whisperId: String) : AuthState()
    data class Error(val message: String) : AuthState()
}

@Singleton
class AuthService @Inject constructor(
    private val wsClient: WsClientImpl,
    private val secureStorage: SecureStorage,
    private val cryptoService: CryptoService,
    private val gson: Gson
) {
    private val _authState = MutableStateFlow<AuthState>(
        if (secureStorage.isLoggedIn()) AuthState.Authenticated(secureStorage.whisperId!!)
        else AuthState.Unauthenticated
    )
    val authState: StateFlow<AuthState> = _authState

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // Pending continuations for auth flow
    private var challengeContinuation: CancellableContinuation<RegisterChallengePayload>? = null
    private var ackContinuation: CancellableContinuation<RegisterAckPayload>? = null

    init {
        // Listen for auth-related messages
        scope.launch {
            wsClient.messages.collect { frame ->
                handleMessage(frame)
            }
        }
    }

    /**
     * Register a new account with the server.
     * Server assigns the WhisperID after verifying the public keys.
     */
    suspend fun registerNewAccount(mnemonic: String): Result<Unit> {
        return try {
            _authState.value = AuthState.Authenticating

            // 1. Derive keys from mnemonic
            val keys = cryptoService.deriveKeys(mnemonic)

            // Store keys securely (needed for signing the challenge)
            secureStorage.mnemonic = mnemonic
            secureStorage.encPublicKey = keys.encPublicKey
            secureStorage.encPrivateKey = keys.encPrivateKey
            secureStorage.signPublicKey = keys.signPublicKey
            secureStorage.signPrivateKey = keys.signPrivateKey
            secureStorage.contactsKey = keys.contactsKey

            // 2. Connect to WebSocket
            wsClient.connect()
            waitForConnection()

            // 3. Perform authentication flow
            val ack = performAuth(
                whisperId = null,  // New registration - server assigns ID
                encPublicKey = keys.encPublicKey,
                signPublicKey = keys.signPublicKey,
                signPrivateKey = keys.signPrivateKey
            )

            // 4. Store server-assigned WhisperID and session
            secureStorage.whisperId = ack.whisperId
            secureStorage.sessionToken = ack.sessionToken
            secureStorage.sessionExpiresAt = ack.sessionExpiresAt

            Logger.auth("Registration successful. WhisperID: ${ack.whisperId}")

            // 5. Update auth state
            _authState.value = AuthState.Authenticated(ack.whisperId)

            Result.success(Unit)
        } catch (e: Exception) {
            Logger.e("Registration failed", e)
            _authState.value = AuthState.Error(e.message ?: "Registration failed")
            Result.failure(e)
        }
    }

    /**
     * Recover an existing account using mnemonic.
     * Server will recognize the public keys and return the existing WhisperID.
     */
    suspend fun recoverAccount(mnemonic: String): Result<Unit> {
        return try {
            _authState.value = AuthState.Authenticating

            // Derive keys from mnemonic
            val keys = cryptoService.deriveKeys(mnemonic)

            // Store keys securely
            secureStorage.mnemonic = mnemonic
            secureStorage.encPublicKey = keys.encPublicKey
            secureStorage.encPrivateKey = keys.encPrivateKey
            secureStorage.signPublicKey = keys.signPublicKey
            secureStorage.signPrivateKey = keys.signPrivateKey
            secureStorage.contactsKey = keys.contactsKey

            // Connect and authenticate
            wsClient.connect()
            waitForConnection()

            // Perform auth - server will return existing WhisperID for these keys
            val ack = performAuth(
                whisperId = null,  // Let server find by public keys
                encPublicKey = keys.encPublicKey,
                signPublicKey = keys.signPublicKey,
                signPrivateKey = keys.signPrivateKey
            )

            // Store session
            secureStorage.whisperId = ack.whisperId
            secureStorage.sessionToken = ack.sessionToken
            secureStorage.sessionExpiresAt = ack.sessionExpiresAt

            Logger.auth("Recovery successful. WhisperID: ${ack.whisperId}")

            _authState.value = AuthState.Authenticated(ack.whisperId)

            Result.success(Unit)
        } catch (e: Exception) {
            Logger.e("Recovery failed", e)
            _authState.value = AuthState.Error(e.message ?: "Recovery failed")
            Result.failure(e)
        }
    }

    /**
     * Reconnect with existing session.
     */
    suspend fun reconnect(): Result<Unit> {
        val encPub = secureStorage.encPublicKey
        val signPub = secureStorage.signPublicKey
        val signPriv = secureStorage.signPrivateKey
        val savedWhisperId = secureStorage.whisperId

        if (encPub == null || signPub == null || signPriv == null || savedWhisperId == null) {
            return Result.failure(AuthException("Not authenticated - missing keys"))
        }

        return try {
            _authState.value = AuthState.Authenticating

            wsClient.connect()
            waitForConnection()

            val ack = performAuth(
                whisperId = savedWhisperId,  // Existing session
                encPublicKey = encPub,
                signPublicKey = signPub,
                signPrivateKey = signPriv
            )

            secureStorage.sessionToken = ack.sessionToken
            secureStorage.sessionExpiresAt = ack.sessionExpiresAt

            Logger.auth("Reconnect successful")

            _authState.value = AuthState.Authenticated(ack.whisperId)

            Result.success(Unit)
        } catch (e: Exception) {
            Logger.e("Reconnect failed", e)
            _authState.value = AuthState.Error(e.message ?: "Reconnect failed")
            Result.failure(e)
        }
    }

    /**
     * Perform the authentication flow:
     * 1. Send register_begin
     * 2. Receive register_challenge
     * 3. Sign challenge
     * 4. Send register_proof
     * 5. Receive register_ack with WhisperID
     */
    private suspend fun performAuth(
        whisperId: String?,
        encPublicKey: ByteArray,
        signPublicKey: ByteArray,
        signPrivateKey: ByteArray
    ): RegisterAckPayload {
        val deviceId = secureStorage.getOrCreateDeviceId()

        // Get FCM token for push notifications
        val fcmToken = getFcmToken()

        // Step 1: Send register_begin
        val beginPayload = RegisterBeginPayload(
            deviceId = deviceId,
            whisperId = whisperId
        )
        wsClient.send(WsFrame(Constants.MsgType.REGISTER_BEGIN, payload = beginPayload))

        // Step 2: Wait for challenge
        val challenge = withTimeout(30_000) {
            suspendCancellableCoroutine { cont ->
                challengeContinuation = cont
            }
        }

        // Step 3: Sign challenge - server expects SHA256(challengeBytes) then sign
        val challengeBytes = Base64.decode(challenge.challenge, Base64.NO_WRAP)
        val signature = cryptoService.signChallenge(challengeBytes, signPrivateKey)

        // Step 4: Send register_proof with FCM token
        val proofPayload = RegisterProofPayload(
            challengeId = challenge.challengeId,
            deviceId = deviceId,
            whisperId = whisperId,
            encPublicKey = Base64.encodeToString(encPublicKey, Base64.NO_WRAP),
            signPublicKey = Base64.encodeToString(signPublicKey, Base64.NO_WRAP),
            signature = Base64.encodeToString(signature, Base64.NO_WRAP),
            pushToken = fcmToken
        )
        wsClient.send(WsFrame(Constants.MsgType.REGISTER_PROOF, payload = proofPayload))

        // Step 5: Wait for ack
        return withTimeout(30_000) {
            suspendCancellableCoroutine { cont ->
                ackContinuation = cont
            }
        }
    }

    /**
     * Get FCM token for push notifications.
     * Returns cached token or fetches fresh one from Firebase.
     */
    private suspend fun getFcmToken(): String? {
        return try {
            // Try cached token first
            secureStorage.fcmToken?.let { return it }

            // Fetch from Firebase
            val token = FirebaseMessaging.getInstance().token.await()
            secureStorage.fcmToken = token
            Logger.auth("FCM token obtained: ${token.take(20)}...")
            token
        } catch (e: Exception) {
            Logger.e("Failed to get FCM token", e)
            null
        }
    }

    /**
     * Sync FCM token with server after successful authentication.
     * This ensures the server has the latest token even after reinstall.
     */
    private fun syncFcmTokenAfterAuth() {
        scope.launch {
            try {
                // Force get fresh token from Firebase
                val token = FirebaseMessaging.getInstance().token.await()
                val sessionToken = secureStorage.sessionToken ?: return@launch

                // Update local cache
                secureStorage.fcmToken = token

                // Send to server
                val payload = UpdateTokensPayload(
                    sessionToken = sessionToken,
                    pushToken = token
                )
                wsClient.send(WsFrame(Constants.MsgType.UPDATE_TOKENS, payload = payload))
                Logger.auth("FCM token synced after auth: ${token.take(20)}...")
            } catch (e: Exception) {
                Logger.e("Failed to sync FCM token after auth", e)
            }
        }
    }

    private suspend fun waitForConnection() {
        var attempts = 0
        while (wsClient.connectionState.value != WsConnectionState.CONNECTED && attempts < 100) {
            delay(100)
            attempts++
        }
        if (wsClient.connectionState.value != WsConnectionState.CONNECTED) {
            throw AuthException("Failed to connect to server")
        }
    }

    private fun handleMessage(frame: WsFrame<JsonElement>) {
        when (frame.type) {
            Constants.MsgType.REGISTER_CHALLENGE -> {
                try {
                    val payload = gson.fromJson(frame.payload, RegisterChallengePayload::class.java)
                    Logger.auth("Received challenge: ${payload.challengeId}")
                    challengeContinuation?.resume(payload)
                    challengeContinuation = null
                } catch (e: Exception) {
                    Logger.e("Failed to parse challenge", e)
                    challengeContinuation?.resumeWithException(e)
                    challengeContinuation = null
                }
            }

            Constants.MsgType.REGISTER_ACK -> {
                try {
                    val payload = gson.fromJson(frame.payload, RegisterAckPayload::class.java)
                    Logger.auth("Received ack: ${payload.whisperId}, success=${payload.success}")
                    if (payload.success) {
                        // Sync FCM token after successful auth
                        syncFcmTokenAfterAuth()
                        ackContinuation?.resume(payload)
                    } else {
                        ackContinuation?.resumeWithException(AuthException("Registration rejected"))
                    }
                    ackContinuation = null
                } catch (e: Exception) {
                    Logger.e("Failed to parse ack", e)
                    ackContinuation?.resumeWithException(e)
                    ackContinuation = null
                }
            }

            Constants.MsgType.ERROR -> {
                try {
                    val payload = gson.fromJson(frame.payload, ErrorPayload::class.java)
                    Logger.e("Auth error: ${payload.code} - ${payload.message}")
                    val exception = AuthException("${payload.code}: ${payload.message}")
                    challengeContinuation?.resumeWithException(exception)
                    ackContinuation?.resumeWithException(exception)
                    challengeContinuation = null
                    ackContinuation = null
                } catch (e: Exception) {
                    Logger.e("Failed to parse error", e)
                }
            }

            Constants.MsgType.SESSION_REFRESH_ACK -> {
                try {
                    val payload = gson.fromJson(frame.payload, SessionRefreshAckPayload::class.java)
                    Logger.auth("Session refreshed: expires at ${payload.sessionExpiresAt}")
                    secureStorage.sessionToken = payload.sessionToken
                    secureStorage.sessionExpiresAt = payload.sessionExpiresAt
                } catch (e: Exception) {
                    Logger.e("Failed to parse session refresh ack", e)
                }
            }
        }
    }

    fun logout() {
        scope.launch {
            try {
                val token = secureStorage.sessionToken
                if (token != null) {
                    // Send logout to server
                    val payload = LogoutPayload(sessionToken = token)
                    wsClient.send(WsFrame(
                        type = Constants.MsgType.LOGOUT,
                        payload = payload
                    ))
                }
            } catch (e: Exception) {
                Logger.e("Logout send failed", e)
            }
        }

        wsClient.disconnect()
        secureStorage.clearAll()
        _authState.value = AuthState.Unauthenticated
    }
}

class AuthException(message: String) : Exception(message)

File: ios/Services/AuthService.swift
import Foundation
import Combine

/// Authentication service - matches server AuthService.ts
final class AuthService: ObservableObject {
    static let shared = AuthService()
    
    @Published private(set) var isAuthenticated = false
    @Published private(set) var whisperId: String?
    @Published private(set) var sessionToken: String?
    
    private let keychain = KeychainService.shared
    private let crypto = CryptoService.shared
    private let ws = WebSocketService.shared
    private var cancellables = Set<AnyCancellable>()
    
    // Keys in memory
    private var encKeyPair: CryptoService.KeyPair?
    private var signKeyPair: CryptoService.KeyPair?
    
    private var pendingChallenge: (id: String, bytes: Data)?
    private var authContinuation: CheckedContinuation<RegisterAckPayload, Error>?
    private var isAuthenticating = false  // Prevent duplicate auth attempts
    private let authLock = NSLock()  // Thread-safe auth state

    // Track which "connection instance" we authenticated on
    // Reset when connection is lost, so we know to re-auth on new connection
    private var authenticatedConnectionId: UUID?
    private var currentConnectionId: UUID?

    private init() {
        setupMessageHandler()
        setupReconnectHandler()
    }

    /// Handle WebSocket reconnections - need to re-authenticate
    private func setupReconnectHandler() {
        ws.$connectionState
            .sink { [weak self] state in
                guard let self = self else { return }

                switch state {
                case .disconnected, .reconnecting:
                    // Reset authenticated state - server requires new auth on each connection
                    if self.isAuthenticated {
                        print("WebSocket disconnected - resetting authentication state")
                        DispatchQueue.main.async {
                            self.isAuthenticated = false
                        }
                    }
                    // Clear connection tracking - this connection is no longer valid
                    self.authLock.lock()
                    self.currentConnectionId = nil
                    self.authenticatedConnectionId = nil
                    self.isAuthenticating = false
                    // Cancel any pending auth continuation
                    if let cont = self.authContinuation {
                        self.authContinuation = nil
                        self.authLock.unlock()
                        cont.resume(throwing: AuthError.registrationFailed("Connection lost"))
                    } else {
                        self.authLock.unlock()
                    }

                case .connected:
                    // Generate a new connection ID for this connection
                    let newConnectionId = UUID()
                    self.authLock.lock()
                    self.currentConnectionId = newConnectionId
                    // Check if we need to authenticate on this NEW connection
                    // Use connection ID instead of isAuthenticated to avoid race conditions
                    let needsAuth = self.authenticatedConnectionId != newConnectionId &&
                                    !self.isAuthenticating &&
                                    self.keychain.whisperId != nil
                    self.authLock.unlock()

                    if needsAuth {
                        print("WebSocket connected (connId: \(newConnectionId.uuidString.prefix(8))) - authenticating...")
                        Task {
                            do {
                                try await self.reconnect()
                            } catch {
                                print("Re-authentication failed: \(error)")
                            }
                        }
                    } else {
                        print("WebSocket connected - already authenticated or auth in progress")
                    }

                case .connecting:
                    break
                }
            }
            .store(in: &cancellables)
    }
    
    /// Current user with keys and session info
    var currentUser: LocalUser? {
        guard let whisperId = whisperId,
              let encKP = encKeyPair,
              let signKP = signKeyPair else {
            return nil
        }
        return LocalUser(
            whisperId: whisperId,
            encPublicKey: encKP.publicKey,
            encPrivateKey: encKP.privateKey,
            signPublicKey: signKP.publicKey,
            signPrivateKey: signKP.privateKey,
            seedPhrase: keychain.mnemonic ?? "",
            sessionToken: sessionToken,
            sessionExpiresAt: nil,
            deviceId: keychain.getOrCreateDeviceId()
        )
    }
    
    private func setupMessageHandler() {
        ws.messagePublisher
            .sink { [weak self] data in
                self?.handleMessage(data)
            }
            .store(in: &cancellables)
    }
    
    // MARK: - Register New Account
    
    func registerNewAccount(mnemonic: String) async throws {
        print("[AuthService] Starting new account registration...")

        // 1. Derive keys from mnemonic
        print("[AuthService] Step 1: Deriving keys from mnemonic...")
        print("[AuthService] Mnemonic word count: \(mnemonic.split(separator: " ").count)")
        print("[AuthService] Mnemonic first word: \(mnemonic.split(separator: " ").first ?? "none")")
        print("[AuthService] Mnemonic last word: \(mnemonic.split(separator: " ").last ?? "none")")
        let derivedKeys = try KeyDerivation.deriveKeys(from: mnemonic)
        print("[AuthService] Keys derived successfully")
        print("[AuthService] encSeed hash: \(derivedKeys.encSeed.prefix(4).map { String(format: "%02x", $0) }.joined())")
        print("[AuthService] signSeed hash: \(derivedKeys.signSeed.prefix(4).map { String(format: "%02x", $0) }.joined())")

        // 2. Generate key pairs
        print("[AuthService] Step 2: Generating key pairs...")
        let encKP = try crypto.generateEncryptionKeyPair(from: derivedKeys.encSeed)
        let signKP = try crypto.generateSigningKeyPair(from: derivedKeys.signSeed)
        print("[AuthService] Key pairs generated")
        print("[AuthService] encPublicKey: \(encKP.publicKey.prefix(8).map { String(format: "%02x", $0) }.joined())...")
        print("[AuthService] signPublicKey: \(signKP.publicKey.prefix(8).map { String(format: "%02x", $0) }.joined())...")

        // 3. Store in keychain
        print("[AuthService] Step 3: Storing keys in keychain...")
        try keychain.storeKeys(
            encPrivateKey: encKP.privateKey,
            encPublicKey: encKP.publicKey,
            signPrivateKey: signKP.privateKey,
            signPublicKey: signKP.publicKey,
            contactsKey: derivedKeys.contactsKey
        )
        keychain.mnemonic = mnemonic
        print("[AuthService] Keys stored in keychain")

        // 4. Keep in memory
        self.encKeyPair = encKP
        self.signKeyPair = signKP

        // 5. Connect and authenticate
        print("[AuthService] Step 5: Connecting to WebSocket...")
        ws.connect()
        try await waitForConnection()
        print("[AuthService] WebSocket connected")

        // 6. Perform auth flow
        print("[AuthService] Step 6: Performing authentication...")
        let ack = try await performAuth(
            whisperId: nil,  // New registration - server assigns ID
            encPublicKey: encKP.publicKey,
            signPublicKey: signKP.publicKey,
            signPrivateKey: signKP.privateKey
        )
        print("[AuthService] Authentication complete, received whisperId: \(ack.whisperId)")

        // 7. Store session and mark connection as authenticated
        print("[AuthService] Step 7: Storing session...")
        authLock.lock()
        authenticatedConnectionId = currentConnectionId
        authLock.unlock()

        await MainActor.run {
            self.whisperId = ack.whisperId
            self.sessionToken = ack.sessionToken
            self.isAuthenticated = true
        }
        keychain.whisperId = ack.whisperId
        keychain.sessionToken = ack.sessionToken
        print("[AuthService] Session stored, isAuthenticated = true, connId: \(currentConnectionId?.uuidString.prefix(8) ?? "nil")")

        // 8. Send push tokens now that we're authenticated
        print("[AuthService] Step 8: Sending push tokens...")
        await PushNotificationService.shared.sendTokensAfterAuth()

        // 9. Fetch any pending messages from server
        print("[AuthService] Step 9: Fetching pending messages...")
        await fetchPendingMessagesAfterAuth()
        print("[AuthService] Registration complete!")
    }

    // MARK: - Recover Account
    
    func recoverAccount(mnemonic: String) async throws {
        guard KeyDerivation.isValidMnemonic(mnemonic) else {
            throw CryptoError.invalidMnemonic
        }
        
        // Same as register but whisperId may be provided for recovery
        // For now, treat same as new (server matches by public keys)
        try await registerNewAccount(mnemonic: mnemonic)
    }
    
    // MARK: - Reconnect with existing session

    func reconnect() async throws {
        // Prevent duplicate auth attempts with thread-safe check
        authLock.lock()
        guard !isAuthenticating else {
            authLock.unlock()
            print("Authentication already in progress, skipping")
            return
        }
        isAuthenticating = true
        authLock.unlock()

        defer {
            authLock.lock()
            isAuthenticating = false
            authLock.unlock()
        }

        guard let encPriv = keychain.getData(forKey: Constants.StorageKey.encPrivateKey),
              let encPub = keychain.getData(forKey: Constants.StorageKey.encPublicKey),
              let signPriv = keychain.getData(forKey: Constants.StorageKey.signPrivateKey),
              let signPub = keychain.getData(forKey: Constants.StorageKey.signPublicKey),
              let savedWhisperId = keychain.whisperId else {
            throw AuthError.notAuthenticated
        }

        self.encKeyPair = CryptoService.KeyPair(publicKey: encPub, privateKey: encPriv)
        self.signKeyPair = CryptoService.KeyPair(publicKey: signPub, privateKey: signPriv)

        // Only connect if not already connected
        if ws.connectionState != .connected {
            ws.connect()
            try await waitForConnection()
        }

        let ack = try await performAuth(
            whisperId: savedWhisperId,  // Recovery
            encPublicKey: encPub,
            signPublicKey: signPub,
            signPrivateKey: signPriv
        )

        // Mark this connection as authenticated
        authLock.lock()
        authenticatedConnectionId = currentConnectionId
        authLock.unlock()

        await MainActor.run {
            self.whisperId = ack.whisperId
            self.sessionToken = ack.sessionToken
            self.isAuthenticated = true
        }
        keychain.sessionToken = ack.sessionToken

        print("Authentication complete on connection \(currentConnectionId?.uuidString.prefix(8) ?? "nil")")

        // Send push tokens now that we're authenticated
        await PushNotificationService.shared.sendTokensAfterAuth()

        // Fetch any pending messages from server
        await fetchPendingMessagesAfterAuth()
    }

    /// Fetch pending messages after authentication
    private func fetchPendingMessagesAfterAuth() async {
        do {
            print("Fetching pending messages after authentication...")
            try await MessagingService.shared.fetchPendingMessages()
            print("Pending messages fetch requested")
        } catch {
            print("Failed to fetch pending messages: \(error)")
        }
    }

    // MARK: - Logout
    
    func logout() {
        if let token = sessionToken {
            let payload = LogoutPayload(sessionToken: token)
            let frame = WsFrame(type: Constants.MessageType.logout, payload: payload)
            Task { try? await ws.send(frame) }
        }
        
        ws.disconnect()
        keychain.clearAll()
        
        encKeyPair = nil
        signKeyPair = nil
        whisperId = nil
        sessionToken = nil
        isAuthenticated = false
    }
    
    // MARK: - Private
    
    private func waitForConnection() async throws {
        for _ in 0..<50 {
            if ws.connectionState == .connected {
                return
            }
            try await Task.sleep(nanoseconds: 100_000_000)
        }
        throw NetworkError.timeout
    }
    
    private func performAuth(
        whisperId: String?,
        encPublicKey: Data,
        signPublicKey: Data,
        signPrivateKey: Data
    ) async throws -> RegisterAckPayload {
        let deviceId = keychain.getOrCreateDeviceId()

        // Cancel any existing continuation to prevent leaks
        authLock.lock()
        if let existingContinuation = authContinuation {
            authContinuation = nil
            authLock.unlock()
            existingContinuation.resume(throwing: AuthError.registrationFailed("Auth cancelled - new attempt started"))
        } else {
            authLock.unlock()
        }

        // Clear any stale challenge
        pendingChallenge = nil

        // Step 1: Send register_begin
        let beginPayload = RegisterBeginPayload(deviceId: deviceId, whisperId: whisperId)
        let beginFrame = WsFrame(type: Constants.MessageType.registerBegin, payload: beginPayload)
        try await ws.send(beginFrame)

        // Wait for challenge
        let challenge = try await waitForChallenge()

        // Step 2: Sign challenge - SHA256(challengeBytes) then sign
        let challengeBytes = Data(base64Encoded: challenge.challenge)!
        let signature = try crypto.signChallenge(challengeBytes, privateKey: signPrivateKey)

        // Step 3: Send register_proof
        let proofPayload = RegisterProofPayload(
            challengeId: challenge.challengeId,
            deviceId: deviceId,
            whisperId: whisperId,
            encPublicKey: encPublicKey.base64EncodedString(),
            signPublicKey: signPublicKey.base64EncodedString(),
            signature: signature.base64EncodedString()
        )
        let proofFrame = WsFrame(type: Constants.MessageType.registerProof, payload: proofPayload)
        try await ws.send(proofFrame)

        // Wait for ack with timeout
        return try await withCheckedThrowingContinuation { continuation in
            authLock.lock()
            self.authContinuation = continuation
            authLock.unlock()

            // Set a timeout to prevent hanging forever
            Task {
                try? await Task.sleep(nanoseconds: 10_000_000_000) // 10 seconds
                self.authLock.lock()
                if let cont = self.authContinuation {
                    self.authContinuation = nil
                    self.authLock.unlock()
                    cont.resume(throwing: AuthError.registrationFailed("Auth timeout"))
                } else {
                    self.authLock.unlock()
                }
            }
        }
    }
    
    private func waitForChallenge() async throws -> RegisterChallengePayload {
        for _ in 0..<100 {
            if let challenge = pendingChallenge {
                pendingChallenge = nil
                return RegisterChallengePayload(
                    challengeId: challenge.id,
                    challenge: challenge.bytes.base64EncodedString(),
                    expiresAt: Int64(Date().timeIntervalSince1970 * 1000) + 60000
                )
            }
            try await Task.sleep(nanoseconds: 50_000_000)
        }
        throw AuthError.invalidChallenge
    }
    
    private func handleMessage(_ data: Data) {
        guard let raw = try? JSONDecoder().decode(RawWsFrame.self, from: data) else { return }
        
        switch raw.type {
        case Constants.MessageType.registerChallenge:
            if let frame = try? JSONDecoder().decode(WsFrame<RegisterChallengePayload>.self, from: data),
               let bytes = Data(base64Encoded: frame.payload.challenge) {
                pendingChallenge = (frame.payload.challengeId, bytes)
            }
            
        case Constants.MessageType.registerAck:
            if let frame = try? JSONDecoder().decode(WsFrame<RegisterAckPayload>.self, from: data) {
                authLock.lock()
                let cont = authContinuation
                authContinuation = nil
                authLock.unlock()
                cont?.resume(returning: frame.payload)
            }

        case Constants.MessageType.error:
            if let frame = try? JSONDecoder().decode(WsFrame<ErrorPayload>.self, from: data) {
                authLock.lock()
                let cont = authContinuation
                authContinuation = nil
                authLock.unlock()
                cont?.resume(throwing: AuthError.registrationFailed(frame.payload.message))
            }

        case Constants.MessageType.sessionRefreshAck:
            if let frame = try? JSONDecoder().decode(WsFrame<SessionRefreshAckPayload>.self, from: data) {
                // Update stored session token and expiry
                keychain.setString(frame.payload.sessionToken, forKey: Constants.StorageKey.sessionToken)
                print("[AuthService] Session refreshed: expires at \(frame.payload.sessionExpiresAt)")
            }

        default:
            break
        }
    }
}
